src/miniscript/musig_key.rs

```diff
@@ -0,0 +1,166 @@
+//! Support for multi-signature keys
+use core::fmt;
+use core::str::FromStr;
+
+use crate::expression::{FromTree, Tree};
+use crate::prelude::*;
+use crate::{Error, MiniscriptKey};
+
+#[derive(Clone, PartialEq, PartialOrd, Ord, Eq, Hash)]
+/// Enum for representing musig keys in miniscript
+pub enum KeyExpr<Pk: MiniscriptKey> {
+    /// Single-key (e.g pk(a), here 'a' is a single key)
+    SingleKey(Pk),
+
+    /// Collection of keys in used for musig-signature
+    MuSig(Vec<KeyExpr<Pk>>),
+}
+
+impl<Pk: MiniscriptKey + FromStr> FromTree for KeyExpr<Pk> {
+    fn from_tree(tree: &Tree) -> Result<KeyExpr<Pk>, Error> {
+        if tree.name == "musig" {
+            let mut key_expr_vec = vec![];
+            for sub_tree in tree.args.iter() {
+                let temp_res = KeyExpr::<Pk>::from_tree(sub_tree)?;
+                key_expr_vec.push(temp_res);
+            }
+            Ok(KeyExpr::MuSig(key_expr_vec))
```

Comment:
nit: Can be re-writtten as:
```suggestion
            let key_expr_vec = tree
                .args
                .iter()
                .map(|subtree| KeyExpr::<Pk>::from_tree(subtree))
                .collect::<Result<Vec<KeyExpr<Pk>>, Error>>()?;
            Ok(KeyExpr::MuSig(key_expr_vec))

```
It's not the cleanest solution so really upto you if you wanna include this :p

Reply:
Looks better and more readable. Changed according to the suggestion.